src/webgpu/shader/execution/expression/expression.ts

```diff
@@ -664,41 +667,53 @@ function basicExpressionShaderBody(
     uniqueID: () => `cts_symbol_${nextUniqueIDSuffix++}`,
   };
   if (inputSource === 'const') {
+    let constEvaluationMode = params.constEvaluationMode;
+    if (constEvaluationMode === undefined) {
+      if (parameterTypes.some(ty => isAbstractType(scalarTypeOf(ty)))) {
+        // Directly assign the expression to the output, to avoid an
+        // intermediate store, which will concretize the value early
```

Comment:
Are there valid use cases for non-'direct' modes with abstract types? If no, add an assert to make sure no one accidentally triggers this early concretization. If yes, maybe consider names like "unrolled-with-early-concretization"..?